packages/mcfunction/src/parser/macro.ts

```diff
@@ -0,0 +1,153 @@
+import * as core from '@spyglassmc/core'
+import { localeQuote, localize } from '@spyglassmc/locales'
+import type { MacroChildNode, MacroKeyNode, MacroNode } from '../node/index.js'
+
+/**
+ * Parse a macro.
+ */
+export function macro(): core.Parser<MacroNode> {
+	return (
+		src: core.Source,
+		ctx: core.ParserContext,
+	): core.Result<MacroNode> => {
+		const start = src.cursor
+		const children: MacroChildNode[] = []
+
+		// Handle the starting '$'
+		if (src.peek() == '$' && src.peek(2) !== '$(') {
+			const ans: MacroChildNode = {
+				type: 'mcfunction:macro_child',
+				range: core.Range.create(start, start + 1),
+				options: { type: 'sign' },
+				value: '$',
+			}
+			src.skip()
+			children.push(ans)
+		}
+
+		// Handle the rest of the line
+
+		let beginning = src.cursor
+		let txt = src.readUntil('$', '\r', '\n')
+		do { // Do-while because we always want it to work the first time
+			let wasMacro = false
```

Comment:
I'll be honest, the `txt` and `wasMacro` variables storing random state make the logic very confusing to follow.

Reply:
You renamed the variables but imo the issue is still there.

Reply:
Tried to explain things some more and reorganize things a bit